[REVIEW] Add matrix gather prim

akkamesh opened the pull request:
Adding functionality and test cases for matrix gather. This functionality can be used to conditionally copy rows from a source matrix into a destination matrix according to a map. Its immediate use is in kmeans.

Tagging @teju85 for review.

GPUtester commented:
Can one of the admins verify this patch?

cjnolet commented:
add to whitelist

teju85 commented on ml-prims/src/matrix/gather.h:
```diff
@@ -0,0 +1,295 @@
+      
+  // return type of MapTransformOp, must be convertable to IndexT
+  typedef typename std::result_of<decltype(transform_op)(MapValueT)>::type MapTransformOpReturnT;
+  ASSERT((std::is_convertible<MapTransformOpReturnT, IndexT>::value),
```
this might as well be a `static_assert`, right?

akkamesh replied:
Yeah you are right, I have fixed it!

teju85 commented on ml-prims/src/matrix/gather.h:
```diff
@@ -0,0 +1,295 @@
+
+  // return type of UnaryPredicateOp, must be convertible to bool
+  typedef typename std::result_of<decltype(pred_op)(StencilValueT)>::type PredicateOpReturnT;
+  ASSERT((std::is_convertible<PredicateOpReturnT, bool>::value),
```
same as before. `static_assert`

teju85 commented:
@akkamesh `map_length` is expected to be less than or equal to `N`, right?

teju85 commented:
@akkamesh In general, all of our prims currently expose raw device pointers in their interface. We can certainly debate whether that's the correct thing to do, but atleast in this sense, the interface is uniform today. Any particular reason for you to choose iterator_traits here?

akkamesh commented:
Thanks for the feedback @teju85 , I have fixed static_assert and IndexT usage.
 
> @akkamesh `map_length` is expected to be less than or equal to `N`, right?

It is only if we assume that the map always points to unique values which may not be always the case. Say if we want to duplicate rows from the input matrix OR create a matrix from a vector OR other such use cases where duplicate map values are allowed then we can expect to have `map_length > N`. Its more general this way (similar to thrust::gather but for matrix).


> @akkamesh In general, all of our prims currently expose raw device pointers in their interface. We can certainly debate whether that's the correct thing to do, but atleast in this sense, the interface is uniform today. Any particular reason for you to choose iterator_traits here?

They allow you to operate on different kinds of containers in a generic way. This includes c-style arrays since pointers are also iterators and I believe iterator support is future proof (especially with so much emphasis on C++ standards in RAPIDS). Also, I was primarily following in the direction of cub and thrust which is well established in the community. 
I understand that this deviates from the uniform interface but at some point if we want to support fancy iterators for some ML algorithms then it keeps the option open (which was the case for reduce_rows_by_key and I added that support recently).

akkamesh commented on ml-prims/src/matrix/gather.h:
```diff
@@ -0,0 +1,295 @@
+	  typename IndexT = int>
+__global__
+void gatherKernel(MatrixIteratorT in,
+		  int D,
```
You are right! I have fixed it.
